ext/mysql2/statement.c

```diff
@@ -0,0 +1,442 @@
+
+  return rb_stmt;
+}
+
+/* call-seq: stmt.param_count # => Numeric
+ *
+ * Returns the number of parameters the prepared statement expects.
+ */
+static VALUE param_count(VALUE self) {
+  GET_STATEMENT(self);
+
+  return ULL2NUM(mysql_stmt_param_count(stmt_wrapper->stmt));
+}
+
+/* call-seq: stmt.field_count # => Numeric
+ *
+ * Returns the number of fields the prepared statement returns.
+ */
+static VALUE field_count(VALUE self) {
+  GET_STATEMENT(self);
+
+  return UINT2NUM(mysql_stmt_field_count(stmt_wrapper->stmt));
+}
+
+static void *nogvl_execute(void *ptr) {
+  MYSQL_STMT *stmt = ptr;
+
+  if(mysql_stmt_execute(stmt)) {
+    return (void*)Qfalse;
+  } else {
+    return (void*)Qtrue;
+  }
+}
+
+static void *nogvl_stmt_store_result(void *ptr) {
+  MYSQL_STMT *stmt = ptr;
+
+  if(mysql_stmt_store_result(stmt)) {
+    return (void *)Qfalse;
+  } else {
+    return (void *)Qtrue;
+  }
+}
+
+#define FREE_BINDS                                          \
+  for (i = 0; i < argc; i++) {                              \
+    if (bind_buffers[i].buffer && NIL_P(params_enc[i])) {   \
+      xfree(bind_buffers[i].buffer);                        \
+    }                                                       \
+  }                                                         \
+  if (argc > 0) xfree(bind_buffers);
+
+/* call-seq: stmt.execute
+ *
+ * Executes the current prepared statement, returns +result+.
+ */
+static VALUE execute(int argc, VALUE *argv, VALUE self) {
+  MYSQL_BIND *bind_buffers = NULL;
+  unsigned long bind_count;
+  long i;
+  MYSQL_STMT *stmt;
+  MYSQL_RES *metadata;
+  VALUE resultObj;
+  VALUE *params_enc = alloca(sizeof(VALUE) * argc);
+  unsigned long* length_buffers = NULL;
+  int is_streaming = 0;
+#ifdef HAVE_RUBY_ENCODING_H
+  rb_encoding *conn_enc;
+#endif
+  GET_STATEMENT(self);
+#ifdef HAVE_RUBY_ENCODING_H
+  {
+    GET_CLIENT(stmt_wrapper->client);
+    conn_enc = rb_to_encoding(wrapper->encoding);
+  }
+#endif
+  {
+    VALUE valStreaming = rb_hash_aref(rb_iv_get(stmt_wrapper->client, "@query_options"), sym_stream);
+    if(valStreaming == Qtrue) {
+      is_streaming = 1;
+    }
+  }
+
+  stmt = stmt_wrapper->stmt;
+
+  bind_count = mysql_stmt_param_count(stmt);
+  if (argc != (long)bind_count) {
+    rb_raise(cMysql2Error, "Bind parameter count (%ld) doesn't match number of arguments (%d)", bind_count, argc);
+  }
+
+  // setup any bind variables in the query
+  if (bind_count > 0) {
+    bind_buffers = xcalloc(bind_count, sizeof(MYSQL_BIND));
+    length_buffers = xcalloc(bind_count, sizeof(unsigned long));
+
+    for (i = 0; i < argc; i++) {
+      bind_buffers[i].buffer = NULL;
+      params_enc[i] = Qnil;
+
+      switch (TYPE(argv[i])) {
+        case T_NIL:
+          bind_buffers[i].buffer_type = MYSQL_TYPE_NULL;
+          break;
+        case T_FIXNUM:
+#if SIZEOF_INT < SIZEOF_LONG
+          bind_buffers[i].buffer_type = MYSQL_TYPE_LONGLONG;
+          bind_buffers[i].buffer = malloc(sizeof(long long int));
```

Comment:
Per https://github.com/brianmario/mysql2/pull/591#issuecomment-91651666 let's switch this to xmalloc unless @brianmario says he switched it for a strong reason?